src/core/ddsi/src/ddsi_security_omg.c

```diff
@@ -2155,31 +2155,41 @@ bool q_omg_security_check_remote_writer_permissions(const struct proxy_writer *p
     }
   }
 
-  if (SECURITY_INFO_IS_WRITE_PROTECTED(pwr->c.security_info))
+  if (!SECURITY_INFO_IS_WRITE_PROTECTED(pwr->c.security_info))
+    return true;
+
+  DDS_Security_PermissionsHandle permissions_handle;
+  if ((permissions_handle = get_permissions_handle(pp, pwr->c.proxypp)) == 0)
   {
-    DDS_Security_PermissionsHandle permissions_handle;
+    GVTRACE("Secure remote writer "PGUIDFMT" proxypp does not have permissions handle yet\n", PGUID(pwr->e.guid));
+    return false;
+  }
 
-    if ((permissions_handle = get_permissions_handle(pp, pwr->c.proxypp)) == 0)
-    {
-      GVTRACE("Secure remote writer "PGUIDFMT" proxypp does not have permissions handle yet\n", PGUID(pwr->e.guid));
-      return false;
-    }
+  q_omg_shallow_copy_PublicationBuiltinTopicDataSecure(&publication_data, &pwr->e.guid, pwr->c.xqos, &pwr->c.security_info);
+  bool result = sc->access_control_context->check_remote_datawriter(sc->access_control_context, permissions_handle, (int)domain_id, &publication_data, &exception);
+  if (!result)
+  {
+    if (!is_topic_discovery_protected(pp->sec_attr->permissions_handle, sc->access_control_context, publication_data.topic_name))
```

Comment:
I have one more question: here, as well as a few lines below and in `q_omg_security_check_remote_reader_permissions`, if this function returns true, it silently rejects the reader/writer, else it rejects with an error message. I don't understand why that difference exists.

Reply:
The reason for not logging this is to prevent topic names from being visible in log files, in case discovery protection is enabled for a topic (topic names could be considered as sensitive data in certain cases). In ca2fbe7 I've changed this so that the message is logged in DDS_LC_TRACE when discovery protection is turned on.